test/framework/src/main/java/org/elasticsearch/cli/MockTerminal.java

```diff
@@ -34,60 +64,43 @@ public class MockTerminal extends Terminal {
     // and keep track of the last position which was read. It means that we will hold onto all input
     // setup for the mock terminal during its lifetime, but this is normally a very small amount of data
     // so in reality it will not matter.
-    private final List<String> textInput = new ArrayList<>();
-    private int textIndex = 0;
-    private final List<String> secretInput = new ArrayList<>();
-    private int secretIndex = 0;
-
-    private boolean hasOutputStream = true;
-
-    public MockTerminal() {
-        super("\n"); // always *nix newlines for tests
-    }
-
-    @Override
-    public String readText(String prompt) {
-        if (textIndex >= textInput.size()) {
-            throw new IllegalStateException("No text input configured for prompt [" + prompt + "]");
-        }
-        return textInput.get(textIndex++);
-    }
-
-    @Override
-    public char[] readSecret(String prompt) {
-        if (secretIndex >= secretInput.size()) {
-            throw new IllegalStateException("No secret input configured for prompt [" + prompt + "]");
-        }
-        return secretInput.get(secretIndex++).toCharArray();
-    }
-
-    @Override
-    public PrintWriter getWriter() {
-        return writer;
-    }
-
-    @Override
-    public OutputStream getOutputStream() {
-        return hasOutputStream ? stdoutBuffer : null;
+    private final LazyByteArrayInputStream stdinBuffer;
+
+    private MockTerminal(
+        LazyByteArrayInputStream stdin,
+        ByteArrayOutputStream stdout,
+        ByteArrayOutputStream stderr,
+        boolean supportsBinary
+    ) {
+        super(
+            new InputStreamReader(stdin),
+            newPrintWriter(stdout),
+            newPrintWriter(stderr),
+            supportsBinary ? stdin : null,
+            supportsBinary ? stdout : null
+        );
+        this.stdinBuffer = stdin;
+        this.stdoutBuffer = stdout;
+        this.stderrBuffer = stderr;
     }
 
-    @Override
-    public PrintWriter getErrorWriter() {
-        return errorWriter;
+    private static PrintWriter newPrintWriter(OutputStream out) {
+        return new PrintWriter(new OutputStreamWriter(out, StandardCharsets.UTF_8));
     }
 
-    public void setHasOutputStream(boolean hasOutputStream) {
-        this.hasOutputStream = hasOutputStream;
+    public static MockTerminal create(boolean supportsBinary) {
```

Comment:
Am I correct in reading that we don't yet call this with an argument of `true`? That won't block approval but I'm curious whether there are plans to use a MockTerminal that supports binary in upcoming work.

Reply:
Yes, that is correct. I plan to work on that in a future change, so that we can fix the input reader to not consume past each line of input (leaving binary data that we could read on the stream).